src/control_devices/CCPACSLeadingEdgeDevice.cpp

```diff
@@ -0,0 +1,335 @@
+    double innerTranslationZ = Interpolate(controlParams, innerZTrans, m_currentControlParam);
+    double outerTranslationX = Interpolate(controlParams, outerXTrans, m_currentControlParam);
+    double outerTranslationZ = Interpolate(controlParams, outerZTrans, m_currentControlParam);
+
+    auto wingTrafo       = Wing().GetTransformation().getTransformationMatrix();
+    gp_Pnt innerHingeOld = wingTrafo.Transform(m_hingePoints->inner);
+    gp_Pnt outerHingeOld = wingTrafo.Transform(m_hingePoints->outer);
+
+    // innerTranslationY on hingePoint1 on purpose, maybe consider setting it to zero as default. See CPACS definition on
+    // Path/Step/HingeLineTransformation for more information.
+    gp_Pnt outerHingeNew = outerHingeOld.XYZ() + gp_XYZ(outerTranslationX, innerTranslationY, outerTranslationZ);
+    gp_Pnt innerHingeNew = innerHingeOld.XYZ() + gp_XYZ(innerTranslationX, innerTranslationY, innerTranslationZ);
+
+    // calculating the needed transformations
+    CTiglControlSurfaceTransformation transformation(innerHingeOld, outerHingeOld, innerHingeNew, outerHingeNew,
+                                                     rotation);
+
+    return transformation.getTotalTransformation();
+}
+
+double CCPACSLeadingEdgeDevice::GetMinControlParameter() const
+{
+    return GetPath().GetSteps().GetSteps().front()->GetControlParameter();
+}
+
+double CCPACSLeadingEdgeDevice::GetMaxControlParameter() const
+{
+    return GetPath().GetSteps().GetSteps().back()->GetControlParameter();
+}
+
+double CCPACSLeadingEdgeDevice::GetControlParameter() const
+{
+    return m_currentControlParam;
+}
+
+void CCPACSLeadingEdgeDevice::SetControlParameter(const double param)
+{
+    // clamp currentControlParam to minimum and maximum values
+    double new_param = Clamp(param, GetMinControlParameter(), GetMaxControlParameter());
+
+    if (fabs(new_param - m_currentControlParam) > 1e-6) {
+        m_currentControlParam = new_param;
+        // make sure the wing gets relofted with flaps
+        Invalidate();
+    }
+}
+
+void CCPACSLeadingEdgeDevice::Invalidate()
+{
+    CCPACSWing& wing = Wing();
+    wing.SetBuildFlaps(true);
+    wing.GetConfiguration().AircraftFusingAlgo()->Invalidate();
+}
+
+const CCPACSWingComponentSegment& ComponentSegment(const CCPACSLeadingEdgeDevice& self)
+{
+    return *self.GetParent()->GetParent()->GetParent();
+}
+
+CCPACSWingComponentSegment& ComponentSegment(CCPACSLeadingEdgeDevice& self)
+{
+    return const_cast<CCPACSWingComponentSegment&>(ComponentSegment(const_cast<const CCPACSLeadingEdgeDevice&>(self)));
+}
+
+const CCPACSWing& CCPACSLeadingEdgeDevice::Wing() const
+{
+    return *ComponentSegment(*this).GetParent()->GetParent();
+}
+
+CCPACSWing& CCPACSLeadingEdgeDevice::Wing()
+{
+    return const_cast<CCPACSWing&>(const_cast<const CCPACSLeadingEdgeDevice*>(this)->Wing());
+}
+
+void CCPACSLeadingEdgeDevice::ComputeHingePoints(CCPACSLeadingEdgeDevice::HingePoints& hingePoints) const
```

Comment:
Suggestion for more DRYness

The implementation of this function could move to a ControlDeviceHelper class. It uses the inner hinge point and outer hinge point of the flap, the component segment and CCPACSControlSurfaceBorderXXX.GetEtaLE().